js/src/chat/chat-tools.ts

```diff
@@ -0,0 +1,299 @@
+import { LitElement, html, TemplateResult } from "lit"
+import { property } from "lit/decorators.js"
+import { unsafeHTML } from "lit/directives/unsafe-html.js"
+
+/**
+ * Custom event interface for hiding tool requests
+ */
+declare global {
+  interface GlobalEventHandlersEventMap {
+    "shiny-tool-request-hide": CustomEvent<{ request_id: string }>
+  }
+}
+
+/**
+ * Base class for tool-related card components
+ */
+class ShinyToolCard extends LitElement {
+  @property({ type: String, attribute: "request-id" })
+  requestId!: string
+
+  @property({ type: String })
+  name!: string
+
+  @property({ type: String })
+  title: string = ""
+
+  @property({ type: String })
+  intent?: string
+
+  @property({ type: Boolean, reflect: true })
+  expanded = false
+
+  @property({ type: String })
+  classStatus: string = ""
+
+  @property({ type: String })
+  icon: string = ""
+
+  @property({ type: String })
+  titleTemplate: string = "{title}"
+
+  createRenderRoot(): ShinyToolCard {
+    return this
+  }
+
+  #toggleCollapse(e: Event) {
+    e.preventDefault()
+    this.expanded = !this.expanded
+    this.requestUpdate()
+  }
+
+  protected formatTitle() {
+    let displayTitle = this.title || `${this.name}()`
+    displayTitle = `<span class="tool-title-name">${displayTitle}</span>`
+    displayTitle = this.titleTemplate.replace("{title}", displayTitle)
+    return html`${unsafeHTML(displayTitle)}`
+  }
+
+  protected renderCard(bodyContent: TemplateResult) {
+    const headerId = `tool-header-${this.requestId}`
+    const contentId = `tool-content-${this.requestId}`
+
+    const headerContent = html`
+      <div class="tool-icon ${this.classStatus}">${unsafeHTML(this.icon)}</div>
+      <div class="tool-title ${this.classStatus}">${this.formatTitle()}</div>
+      <div class="tool-spacer"></div>
+      ${this.intent ? html`<div class="tool-intent">${this.intent}</div>` : ""}
+    `
+
+    return html`
+      <div
+        class="shiny-tool-card card bslib-card bslib-mb-spacing html-fill-item html-fill-container m-0"
+        data-bslib-card-init
+        data-require-bs-caller="chat_ui()"
+        data-require-bs-version="5"
+      >
+        <button
+          class="card-header"
+          id="${headerId}"
+          @click="${this.#toggleCollapse}"
+          aria-expanded="${this.expanded}"
+          aria-controls="${contentId}"
+        >
+          ${headerContent}
+          <div class="collapse-indicator">${unsafeHTML(ICONS.plus)}</div>
+        </button>
+        <div
+          class="card-body bslib-gap-spacing html-fill-item html-fill-container${this
```

Comment:
Assuming we don't actually need filling layout or complex gap spacing

```suggestion
          class="card-body r${this
```